nitro/nitro.py

```diff
@@ -1,155 +1,21 @@
-#!/usr/bin/env python3
-
-
-import logging
-import subprocess
-import time
-import threading
-from queue import Queue, Empty
-from concurrent.futures import ThreadPoolExecutor, wait
-
-from nitro.event import NitroEvent
-from nitro.kvm import KVM, VM
-
-
-def find_qemu_pid(vm_name):
-    logging.info('Finding QEMU pid for domain {}'.format(vm_name))
-    libvirt_vm_pid_file = '/var/run/libvirt/qemu/{}.pid'.format(vm_name)
-    try:
-        with open(libvirt_vm_pid_file, 'r') as f:
-            content = f.read()
-            pid = int(content)
-            return pid
-    except IOError:
-        # permission denied
-        # find the PID using pgrep
-        # pgrep --full to match the whole command line
-        cmd = ['pgrep', '--full', 'qemu.*-name.*{}'.format(vm_name)]
-        output = subprocess.check_output(cmd)
-        try:
-            pid = int(output)
-            return pid
-        except ValueError as e:
-            # output issue
-            logging.critical('Cannot find PID with pgrep output: {}'.format(output))
-            raise e
-
+from nitro.listener import Listener
+from nitro.backends import get_backend
 
 class Nitro:
 
-    __slots__ = (
-        'domain',
-        'pid',
-        'kvm_io',
-        'vm_io',
-        'vcpus_io',
-        'stop_request',
-        'futures',
-        'queue',
-        'current_cont_event',
-    )
+    def __init__(self, domain, introspection=True):
+        self.listener = Listener(domain)
+        self.introspection = introspection
+        if self.introspection:
+            self.backend = get_backend(domain)
```

Comment:
I think it would be nicer to set backend to None instead of leaving it out completelly if introspection is not enabled. This would also enable us to leave out the introspection property and just test if backend is None in stop. Of course this is all pretty subjective.